File: init/timesync.c
#include "timesync.h"

void clock_worker()
{
    int sockfd, n;
    struct sockaddr_vm serveraddr;
    char buf[BUFSIZE];
    struct timespec gtime;
    struct timespec htime;
    uint64_t gtime_ns;
    uint64_t htime_ns;

    sockfd = socket(AF_VSOCK, SOCK_DGRAM, 0);
    if (sockfd < 0) {
        perror("Couldn't create timesync socket");
        return;
    }

    bzero((char *)&serveraddr, sizeof(serveraddr));
    serveraddr.svm_family = AF_VSOCK;
    serveraddr.svm_port = TSYNC_PORT;
    serveraddr.svm_cid = 3;

    bzero(buf, BUFSIZE);

    n = bind(sockfd, (struct sockaddr *)&serveraddr, sizeof(serveraddr));
    if (n < 0) {
        printf("Couldn't bind timesync socket\n");
        return;
    }

    while (1) {
        n = recv(sockfd, buf, BUFSIZE, 0);
        if (n < 0) {
            perror("Error in timesync recv");
            return;
        } else if (n != 8) {
            printf("Ignoring bogus timesync packet\n");
            continue;
        }

        htime_ns = *(uint64_t *)&buf[0];
        clock_gettime(CLOCK_REALTIME, &gtime);
        gtime_ns = gtime.tv_sec * NANOS_IN_SECOND;
        gtime_ns += gtime.tv_nsec;

        if (llabs(htime_ns - gtime_ns) > DELTA_SYNC) {
            htime.tv_sec = htime_ns / NANOS_IN_SECOND;
            htime.tv_nsec = htime_ns % NANOS_IN_SECOND;
            clock_settime(CLOCK_REALTIME, &htime);
        }
    }
}

File: init/src/timesync.rs
use std::mem;

const TSYNC_PORT: u32 = 123;
const NANOS_IN_SECOND: u64 = 1_000_000_000;
const DELTA_SYNC: u64 = 100_000_000; // 100ms — don't bother adjusting for smaller drifts

pub fn run() {
    let sock = unsafe { libc::socket(libc::AF_VSOCK, libc::SOCK_DGRAM, 0) };
    if sock < 0 {
        return;
    }

    let mut addr: libc::sockaddr_vm = unsafe { mem::zeroed() };
    addr.svm_family = libc::AF_VSOCK as _;
    addr.svm_port = TSYNC_PORT;
    addr.svm_cid = libc::VMADDR_CID_ANY;

    if unsafe {
        libc::bind(
            sock,
            &addr as *const _ as *const libc::sockaddr,
            mem::size_of_val(&addr) as _,
        )
    } < 0
    {
        unsafe { libc::close(sock) };
        return;
    }

    std::thread::Builder::new()
        .name("timesync".into())
        .spawn(move || loop {
            let mut buf = [0u8; 8];
            let n = unsafe { libc::recv(sock, buf.as_mut_ptr() as *mut _, buf.len(), 0) };
            if n < 0 {
                break;
            }
            if n != 8 {
                continue;
            }

            let host_ns = u64::from_le_bytes(buf);

            let mut guest_ts: libc::timespec = unsafe { mem::zeroed() };
            unsafe { libc::clock_gettime(libc::CLOCK_REALTIME, &mut guest_ts) };
            let guest_ns = guest_ts.tv_sec as u64 * NANOS_IN_SECOND + guest_ts.tv_nsec as u64;

            if host_ns.abs_diff(guest_ns) > DELTA_SYNC {
                let host_ts = libc::timespec {
                    tv_sec: (host_ns / NANOS_IN_SECOND) as libc::time_t,
                    tv_nsec: (host_ns % NANOS_IN_SECOND) as libc::c_long,
                };
                unsafe { libc::clock_settime(libc::CLOCK_REALTIME, &host_ts) };
            }
        })
        .unwrap();
}
